feat(sync): Implement prefix deletion

Frando opened the pull request:
## Description

This PR implements prefix deletion as specified in [Willow](https://willowprotocol.org/specs/data-model)

## Notes & open questions

There's an important implication of prefix deletion in our willow data model that we'll have to be very explicit about in docs and tutorials:
```rust
doc.set_bytes(author, b"foobar", b"important data").await?;
doc.set_bytes(author, b"foo", b"crap").await?;
let x = doc.get(author, b"foobar").await?;
// x == None
```
In explanation: The willow data model defines that adding an entry removes all *older entries that have the new entry's key as their prefix*.

What this means is that apps need to model their keyspace carefully. For example, the current file system import/export demo (#1504) will not work as intended, because of the above. 

The recommended solution is that for usecases that need nested keys without them being affected by prefix deletion is to append a nul byte to those keys.

Quoting @AljoschaMeyer from Discord:
> In defense of prefix-based deletion and its useability: the approach was inspired by regular file-system semantics. If I create `/food/pizza.jpg` and `/food/croissant.jpg`, but later created an empty text file /food, it would delete the previously existing /food directory, and hence everything in it.
> Prefix-based deletion was the simplemost solution we could find to replicate this behavior. In particular, we do not need to introduce the concepts of nested directories and leaf files (or the formal counterpart to file system paths: strings of strings).
> The idea was that anyone who'd create an application layer where users deal with actual filesystems would add a small conversion layer: the user-facing api has (nested) directories, (optionally) containing leaf files. These are converted to willow paths in a way that adds a null-byte at the end of every leaf file (and prefixes every directory with a path separator such as /). Willow then replicates exactly the deletion behavior that the users expect from a file system. 


## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [x] Tests if relevant.

dignifiedquire commented:
One thing we should make sure to either unify or clarify is the difference (or not) between `delete_prefix("foo")` and `set("foo", "")`

Frando commented:
> One thing we should make sure to either unify or clarify is the difference (or not) between `delete_prefix("foo")` and `set("foo", "")`
Yup. They are exactly the same with the current implementation. So should we just note this in the docs?

dignifiedquire commented:
I think having the same behaviour is the right call, just need to makes sure it is tested & explained in the docs.

Frando commented:
I pushed a commit that keeps the behavior, but makes the store return `None` for entries with the empty hash (and skip them in the iterators).

dignifiedquire commented on iroh-sync/src/store/fs.rs:
```diff
@@ -513,6 +518,68 @@ impl crate::ranger::Store<SignedEntry> for StoreInstance {
         Ok(iter.chain(iter2))
     }
+
+    // TODO: Not horrible
```
this should be fixed before merging

Frando replied:
Do you have an idea how to solve this efficiently without moving the complete database structure to a tree based one?

dignifiedquire replied:
I meant avoiding the allocations for now, not the algorithm itself, that will need more structural changes from the work @rklaehn will do

Frando replied:
I removed the allocations.

dignifiedquire commented on iroh-sync/src/sync.rs:
```diff
@@ -733,16 +798,54 @@ pub struct Record {
+
+    /// Return `true` if the entry is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len == 0 && self.hash == Hash::EMPTY
```
we use || in other places, maybe here too?

dignifiedquire replied:
we should probably do a consistency check at some point, that if the hash is `EMPTY` the length is set to `0`

Frando replied:
This is mostly covered already:
* `replica.insert()` with `hash::EMPTY` or `len: 0` returns an error so that you're forced to use `wipe_at_prefix` That is debatable - should we allow it instead?
* `wipe_at_prefix` sets `len: 0` and `hash::EMPTY`
* leaves `insert_remote_entry` for gossiped entries - I'll add a check there and return an error for entries with hash::Empty and not len 0.

Frando replied:
Added the validation on `insert_remote_entry`

dignifiedquire commented on iroh-sync/src/sync.rs:
```diff
@@ -1297,6 +1395,64 @@ mod tests {
+        assert_eq!(store.get_one(myspace.id(), alice.id(), b"foo")?, None);
+
+        Ok(())
+    }
```
we will need some more extenstive tests here, especially for the db part, making sure we don’t delete the wrong things

Frando replied:
Do you have some specific test cases in mind here?

dignifiedquire replied:
mostly around the prefix logic, edge cases etc adding and removing things a lot

dignifiedquire reviewed (commented):
looks good overall, but we also need a test syncing nodes checking deletions match up

Frando commented:
Changes since @dignifiedquire's review:
* Added a test that syncs a prefix wipe
* Added validation so that "invalid empty" entries are rejected (empty hash but len not 0, or len 0 but hash not empty hash)
* Removed allocations from the ParentIterator in `fs` impl

Frando commented:
So I wrote a test, found a bug in the range calculation, and fixed it.
This should still get more tests. But now we at least have one test that caught edge case.
Should we go ahead and merge, and then likely look into some proptests or similar? Or does someone have ideas what we should test manually?

Frando commented:
Last commits have some renames - the bikeshed is over, for now: we settled for `del`.

dignifiedquire commented on iroh/src/commands/sync.rs:
```diff
@@ -197,6 +215,21 @@ impl DocCommands {
+                let doc = get_doc(iroh, env, doc).await?;
+                let author = env.author(author)?;
+                let key = prefix.as_bytes().to_vec();
+                let removed = doc.del(author, key).await?;
```
what about confirmation?

dignifiedquire commented on iroh/Cargo.toml:
```diff
@@ -68,6 +68,7 @@ colored = { version = "2.0.4", optional = true }
 # Examples
 ed25519-dalek = { version = "2.0.0", features = ["serde", "rand_core"], optional = true }
 reflink-copy = { version = "0.1.8", optional = true }
+dialoguer = { version = "0.11.0", default-features = false }
```
should be optional and enabled with the `cli` feature

dignifiedquire commented on iroh/src/commands/sync.rs:
```diff
@@ -197,6 +216,31 @@ impl DocCommands {
+                        fmt_short(author)
+                    );
+                } else {
+                    println!("Aborted.")
```
maybe `Canceled`, Aborted sounds like the whole program got killed 🤷
